test/issues/0217.js

```diff
@@ -5,19 +5,107 @@ var assert = require('assert');
-    'and_another_one(); and_another(); and_another(); and_more_and_more(); ' +
-    'until_the_line_become_too_long();}"');
-  var dump = yaml.dump(doc);
-  assert(Math.max.apply(null, dump.split('\n').map(function (str) { return str.length; })) > 80);
+// Simplistic check for folded style header at the end of the first line.
+function isFolded(s) {
+  return s.search(/^[^\n]*>[\-+]?\n/) !== -1;
+}
+
+// Runs one cycle of dump then load. Also checks that dumped result is folded.
+function loadAfterDump(input) {
+  var output = yaml.dump(input);
+  if (!isFolded(output)) {
+    assert.fail(output, '(first line should end with >-, >, or >+)',
+      'Test cannot continue: folded style was expected');
+  }
+  return yaml.load(output);
+}
+
+
+test('Folding Javascript functions preserves content', function () {
+  // Tests loading a function, then tests dumping and loading.
+  function assertFunctionPreserved(functionString, inputs, expectedOutputs, name) {
+    var f = yaml.load('!<tag:yaml.org,2002:js/function> "' + functionString + '"');
+    assert.equal(typeof f, 'function', name + ' should be loaded as a function');
+
+    assert.deepEqual(inputs.map(f), expectedOutputs,
+      name + ' should be loaded correctly');
+
+    assert.deepEqual(inputs.map(loadAfterDump(f)), expectedOutputs,
+      name + ' should be dumped then loaded correctly');
+  }
+
+  // Backslash-escapes double quotes and newlines.
+  function escapeFnString(s) {
+    return s.replace(/"/g, '\\"').replace(/\n/g, '\\n');
+  }
+
+  var fnFactorial = escapeFnString(
+    'function factorial(start) {\n' +
+    '// Non-indented long line to trigger folding: throw new Error("bad fold"); throw new Error("bad fold");\n' +
+    '  var extra_long_string = "try to trick the dumper into creating a syntax error by folding this string";\n' +
+    '  var extra_long_string1 = "try to trick the dumper into creating a syntax error by folding this string";\n' +
+    'var extra_long_string2 = "this long string is fine to fold because it is not more-indented";\n' +
+    'function fac (n) {\n' +
+      'if (n <= 0) return 1; return n * fac(n-1); // here is a long line that can be safely folded\n' +
+    '}\n' +
+    'return fac(start);\n' +
+    '}\n');
+
+  var fnCollatz = escapeFnString(
+    'function collatz(start) {\n' +
+    '  var longString = "another long more-indented string that will cause a syntax error if folded";\n' +
+    'var result = [];\n' +
+    'function go(n) { result.push(n); return (n === 1) ? result : go(n % 2 === 0  ?  n / 2  :  3 * n + 1); }\n' +
+    'return go(start >= 1 ? Math.floor(start) : 1);\n' +
+    '}');
+
+  var fnRot13 = escapeFnString(
+    // single-line function.
+    // note the "{return" is so the line doesn't start with a space.
+    'function rot13(s) {return String.fromCharCode.apply(null, s.split("")' +
+    '.map(function (c) { return ((c.toLowerCase().charCodeAt(0) - 97) + 13) % 26 + 97; })); }'
+  );
```

Comment:
I think these functions and constants should be defined at the top-level as two other helper functions.